File: data-structures/stack/stack.ts
type SNode<T> = {
  value: T;
  next?: SNode<T>;
};

class Stack<T> {
  public length: number;
  private head?: SNode<T>;
  constructor() {
    this.head = undefined;
    this.length = 0;
  }

  push(item: T): void {
    const node: SNode<T> = { value: item };
    if (!this.head) {
      this.head = node;
    } else {
      node.next = this.head;
      this.head = node;
    }
    this.length++;
  }

  pop(): T | undefined {
    if (!this.head) return undefined;
    else {
      const removed = this.head;
      if (this.length === 1) {
        this.head = undefined;
        this.length--;
        return removed.value;
      }
      this.head = removed.next;
      this.length--;
      return removed.value;
    }
  }

  peek(): T | undefined {
    return this.head?.value;
  }
}

File: data-structures/stack/stack.py
from __future__ import annotations
from typing import Optional, TypeVar, Generic

T = TypeVar("T")

class Node(Generic[T]):
    def __init__(self, value: T, next: Optional["Node[T]"] = None):
        self.value: T = value
        self.next: Optional["Node[T]"] = next

class Stack(Generic[T]):
    def __init__(self) -> None:
        self.head: Optional[Node[T]] = None
        self.length: int = 0

    def push(self, item: T) -> None:
        self.head = Node(item, self.head)
        self.length += 1

    def pop(self) -> Optional[T]:
        if self.head is None:
            return None

        removed = self.head
        self.head = removed.next
        self.length -= 1
        return removed.value

    def peek(self) -> Optional[T]:
        return self.head.value if self.head else None

